### ISSUE TITLE
Windows: songs.tmp not renamed after library auto-save

### ISSUE THREAD

**Issue by lazka** [quodlibet/quodlibet#1230]
Original [issue 1230](https://code.google.com/p/quodlibet/issues/detail?id=1230) created by thuerrschmidt on 2013-08-28T21:03:23.000Z:

I installed and set up Quod Libet 2.6.2 on a Windows 7 machine. Everything seemed to work fine at first, just like on my main Ubuntu machine. However, when I closed Quod Libet and started it for the second time, all the songs I had previously imported were gone. My library was empty.

Closer analysis showed that the &quot;songs&quot; file was missing in &quot;%APPDATA%/Quod Libet&quot;. There was, however, a &quot;songs.tmp&quot; file in the same directory. I renamed it to &quot;songs&quot;, restarted Quod Libet, and all my music was back. After a while, with Quod Libet still running, a new &quot;songs.tmp&quot; file appeared and the &quot;songs&quot; file disappeard again. Dal segno al coda ad infinitum.

Running Quod Libet in debug mode, I noticed repeated &quot;Couldn't save library to path&quot; warnings in the output log.

I then went on to set up Quod Libet to run from source (not an easy thing to do in Windows). The warnings seemed to orginate from the &quot;dump_items&quot; function in &quot;quodlibet/library/libraries.py&quot; (I'm not a python guy, and only a very occasional programmer, so please forgive me if I don't get all the details right). There, under &quot;except EnvironmentError:&quot;, I copied and pasted a &quot;util.print_exc()&quot; statement, which gave me this additional information:

WindowsError: [Error 32] The process cannot access the file because it is being used by another process

So it seems &quot;songs.tmp&quot; can't be renamed to &quot;songs&quot; because there's a handle still open on it. Why's that? My guess is that's because the fileobj from further above won't be closed after being flushed and fsync'ed. And that's because the renaming happens within the scope of the corresponding &quot;with&quot; block.

In other words, it's a simple indentation error. When everything from the &quot;No atomic rename on windows&quot; comment onwards until the end of the &quot;dump_items&quot; function is de-indented one level, the fileobj handle on &quot;songs.tmp&quot; will be auto-closed on going out of scope, and the renaming to &quot;songs&quot; will succeed.

&quot;No atomic rename on windows&quot; also seems to be the reason why this bug doesn't affect Linux.

I should probably submit a patch, but I don't know how to do this properly. If more information is needed, I'll certainly do what I can.
